### PR TITLE
perf: lazily create RootCertStore

### PR DESCRIPTION
**Pull request by dsherret**
Depends on https://github.com/denoland/deno/pull/18900

This is about 2% faster for a simple hello world.

### CODE REVIEW

File: cli/args/mod.rs [hunk 1]
```diff
@@ -401,13 +404,62 @@ fn discover_package_json(
...
+}
+
 /// Create and populate a root cert store based on the passed options and
 /// environment.
 pub fn get_root_cert_store(
   maybe_root_path: Option<PathBuf>,
   maybe_ca_stores: Option<Vec<String>>,
   maybe_ca_data: Option<CaData>,
-) -> Result<RootCertStore, AnyError> {
+) -> Result<RootCertStore, RootCertStoreLoadError> {
```
Comment (dsherret) [inline cli/args/mod.rs:462]: I initially thought I would have to keep around the error, so I turned it into a structured type. Then after I realized I didn't realize need this, but kept it anyway.

File: cli/http_util.rs [hunk 2]
```diff
@@ -217,34 +218,68 @@ impl CacheSemantics {
...
 
-#[derive(Debug, Clone)]
-pub struct HttpClient(reqwest::Client);
+pub struct HttpClient {
+  root_cert_store_provider: Option<Arc<dyn RootCertStoreProvider>>,
+  unsafely_ignore_certificate_errors: Option<Vec<String>>,
+  cell: once_cell::sync::OnceCell<reqwest::Client>,
+}
+
+impl std::fmt::Debug for HttpClient {
```
Comment (dsherret) [inline cli/http_util.rs:227]: Should I make `RootCertStoreProvider` require `Debug` instead?

File: ext/fetch/lib.rs [hunk 3]
```diff
@@ -189,6 +191,26 @@ pub struct FetchReturn {
   cancel_handle_rid: Option<ResourceId>,
 }
 
+pub fn get_or_create_client_from_state(
```
Comment (dsherret) [inline ext/fetch/lib.rs:194]: Let me know if you can think of any better alternatives to having this function. This didn't seem too bad, but at the same time not that great.

File: ext/net/ops_tls.rs [hunk 4]
```diff
@@ -813,14 +813,10 @@ where
     .try_borrow::<UnsafelyIgnoreCertificateErrors>()
     .and_then(|it| it.0.clone());
 
-  // TODO(@justinmchase): Ideally the certificate store is created once
-  // and not cloned. The store should be wrapped in Arc<T> to reduce
-  // copying memory unnecessarily.
```
Comment (dsherret) [inline ext/net/ops_tls.rs:818]: I tried wrapping the RootCertStore in an `Arc`, but some of the other code we don't control that this is passed to requires a `RootCertStore`.

File: cli/args/mod.rs (hunk 1 above)
Comment (bartlomieju) [reply cli/args/mod.rs:462]: That sounds good

File: cli/http_util.rs (hunk 2 above)
Comment (bartlomieju) [reply cli/http_util.rs:227]: Either works for me 👍

File: ext/fetch/lib.rs (hunk 3 above)
Comment (bartlomieju) [reply ext/fetch/lib.rs:194]: I think it's fine, communicates intent clearly

File: ext/net/ops_tls.rs (hunk 4 above)
Comment (bartlomieju) [reply ext/net/ops_tls.rs:818]: That's correct and that was my main problem with it. I wonder if we should open a PR to these libraries to accept `Arc<RootCertStore>`

Comment (dsherret) [reply ext/net/ops_tls.rs:818]: Maybe after we get a benchmark.

**Review by bartlomieju — APPROVED** [review]
LGTM, can't wait to see benchmark results. This was a visible contributor to start up speed.

**Pull request merged**